### PR TITLE
Issue 4347 - Tabular view

### LINKED ISSUE #4347: Custom tickets - Tabular view
### Description
Users needs to be able to view tickets outside the viewer and group them by the relevant criteria, so we figured that a table view with groups by is the best to approach this.

### Goals
- [ ] Have a useful view for users to manage tickets in the dashboard.

### Tasks
- [ ] Blank state
- [ ] Templates select
- [ ] Groupby select
- [ ] Logic for grouping by
- [ ] Add the new ticket button (consult with Carmen and George about it)

Nice to have:
- [ ] Routing; a link that selects all the appropiate filters
- [ ] Endpoint for getting the tickets by modelid and template
- [ ] Endpoint for getting the templates by project

### Notes
The initial code for this branch is scaffolding intended to provide the spaces for the final components, change any code as needed.

### Related Resources
[MockUp](https://www.figma.com/file/bRqIl0g9psqD3ISkmxwcNp/3D-Repo?type=design&node-id=410-41485&mode=design)

### PR DESCRIPTION
**Pull request by Amantini1997**
This fixes #4347

#### Description
- Added routing to empty tickets tab (selection)
- Added routing to reach tabular view with template/models/groupBy already set
- Added logic to fetch templates for a project
- Filter by group, by template, by search (only id and title), by completed
- Base logic to edit/add new ticket

### CODE REVIEW

File: frontend/test/projects/projects.sagas.spec.ts
Comment (The-Daniel) [inline frontend/test/projects/projects.sagas.spec.ts:32]: I thought you changed this to 'Projects: sagas' or was that in another PR?

**Review by The-Daniel — CHANGES_REQUESTED** [review]

File: frontend/test/projects/projects.sagas.spec.ts
Comment (Amantini1997) [reply frontend/test/projects/projects.sagas.spec.ts:32]: It was indeed this one, but I forgot to fix it. Will do for the next review

File: frontend/src/v5/store/projects/projects.sagas.ts
Comment (Amantini1997) [reply frontend/src/v5/store/projects/projects.sagas.ts:91]: At the moment, teamspace endpoint would indeed return the same templates as any model, but this is not guaranteed to be the case as it may change in the future. Ideally, we're going to have a project endpoints for templates in the future, but for now, this is the way

File: frontend/src/v5/ui/routes/dashboard/projects/tickets/selectMenus/groupByFormSelect.component.tsx
Comment (Amantini1997) [reply frontend/src/v5/ui/routes/dashboard/projects/tickets/selectMenus/groupByFormSelect.component.tsx:30]: This is temporary design, ideally, that should not be a label, but only a placeholder

File: frontend/src/v5/ui/routes/dashboard/projects/tickets/ticketContent.component.tsx
Comment (Amantini1997) [reply frontend/src/v5/ui/routes/dashboard/projects/tickets/ticketContent.component.tsx:52]: If I get what you mean, that is to have something in the URL to keep track of the selected models, I include them as searchParams

**Review by The-Daniel — APPROVED** [review]

**Comment by sanmont3drepo** [comment]
There are new circular dependencies in this branch!
image

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

File: frontend/src/v5/ui/controls/actionMenu/actionMenu.component.tsx
Comment (Amantini1997) [reply frontend/src/v5/ui/controls/actionMenu/actionMenu.component.tsx:30]: It is used to open the actionMenu based on mouse position at the time of the click as opposed to a point of the component (left-top, right-middle, etc.).

That is being used by only one component so far, that is `TicketsTableGroup`. 
George wanted the action menu on this component 
image
to open where the mouse was as opposed to be always in the middle.

File: frontend/src/v5/ui/controls/chip/chip.types.tsx
Comment (Amantini1997) [reply frontend/src/v5/ui/controls/chip/chip.types.tsx:104]: I saw it was the "unset" state was unreachable, but I actually haven't asked Carmen whether that could be removed. I asked her now, let's see what she replies

**Review by sanmont3drepo — APPROVED** [review]

File: frontend/src/v5/ui/routes/dashboard/projects/tickets/ticketsTable/ticketsTable.component.tsx
Comment (The-Daniel) [inline frontend/src/v5/ui/routes/dashboard/projects/tickets/ticketsTable/ticketsTable.component.tsx:1]: Viewers/commenters should not be able to click the 'create ticket' button.
I believe collaborators can create and edit tickets.
Note that the same should follow for editing tickets, but commenters - I'm assuming - can comment

**Review by The-Daniel — CHANGES_REQUESTED** [review]

File: frontend/src/v5/ui/routes/dashboard/projects/tickets/ticketsTable/ticketsTableContent/ticketsTableGroup/ticketsTableRow/ticketsTableRow.component.tsx
Comment (carmenfan) [reply frontend/src/v5/ui/routes/dashboard/projects/tickets/ticketsTable/ticketsTableContent/ticketsTableGroup/ticketsTableRow/ticketsTableRow.component.tsx:1]: (I swear I'm not being a big brother 😆... ) Saw this on email so I thought I'd reply...

Yes if the template selected doesn't have certain fields they should be removed from the columns.

i.e. assignees, due date etc should only appear if template has has them.

Same goes to safetibase properties.

If the template has them but the field is blank - we should be showing blanks.

File: frontend/src/v5/ui/routes/dashboard/projects/tickets/ticketsTable/newTicketMenu/newTicketMenu.component.tsx
Comment (The-Daniel) [inline frontend/src/v5/ui/routes/dashboard/projects/tickets/ticketsTable/newTicketMenu/newTicketMenu.component.tsx:48]: This only works to stop a 'viewer' from creating an issue if there are multiple models selected. If there is just one model selected then this menu is bypassed and a viewer can start creating a ticket. But this is actually probably just due to the other problem listed below

File: frontend/src/v5/ui/routes/dashboard/projects/tickets/ticketsTable/ticketsTable.component.tsx
Comment (The-Daniel) [inline frontend/src/v5/ui/routes/dashboard/projects/tickets/ticketsTable/ticketsTable.component.tsx:216]: I believe this causes a "Rendered more hooks than during the previous render" crash. I think swapping the conditions around might fix it. (I get this crash when I have a model selected and then I use the input to deselect all models)

The `getSelectedModelsNonViewerPermission().length === 0` part is broke though as it seems to always be true for me no matter what my permissions are. Also the button only appears disabled. It can still be clicked. I think you need another `disabled` prop on the parent NewTicketMenu component to disable the onClick

**Review by The-Daniel — CHANGES_REQUESTED** [review]
image

The columns are still there when they shouldn't be

File: frontend/src/v5/ui/routes/dashboard/projects/tickets/selectMenus/groupByFormSelect.component.tsx
Comment (The-Daniel) [inline frontend/src/v5/ui/routes/dashboard/projects/tickets/selectMenus/groupByFormSelect.component.tsx:30]: This is no longer temporary design so this label should be capitalised

**Review by The-Daniel — CHANGES_REQUESTED** [review]

**Comment by carmenfan** [comment]
@Amantini1997 there's a conflict with staging here 😢

File: frontend/src/v5/ui/routes/dashboard/projects/tickets/ticketsList/slides/ticketSlide.component.tsx
Comment (The-Daniel) [inline frontend/src/v5/ui/routes/dashboard/projects/tickets/ticketsList/slides/ticketSlide.component.tsx:1]: If I have a viewer ticket open in the sidebar (I don't have edit permissions) and then I try to create either a commenter or collaborator permission level ticket the inputs are disabled so I cannot actually create the ticket.
image

**Review by The-Daniel — CHANGES_REQUESTED** [review]

**Review by carmenfan — CHANGES_REQUESTED** [review]
- Can we make it so that it defaults to focus on the title when you go into the new Ticket?
img-16-35-23-04-35

- Leave these fields blank if they are not populated:
image

- Can we make open in 3D A new tab?

- Can we elaborate a bit more in this message? `The selected template has required properties that can only be created in the viewer, please proceed to view this model to create the ticket` - Ideally with `view this model` as a link to the 3D viewer with the model loaded
image

- Can we make it so clicking on the headers will sort ascending/decending base on that column? just like container list (as discussed, if it's too complicated, make it a new issue 😄 
img-16-46-23-04-46

- Some data is lost on the link if the user had to log in
Get the brower to view tickets in tabular view, copy the link and paste it in incognito, log in. The selected federation/containers data is not retained. If you paste it in a browser that is alreday logged in, this information is retained.
img-16-42-23-05-42

**Review by carmenfan — CHANGES_REQUESTED** [review]
#### Group by due date logic isn't quite right
I have 3 tickets [here](https://issue-4347.dev.3drepo.io/v5/dashboard/DevTeamWorkspace/7b6c7bd0-054a-11ee-ac5b-59e952832776/t/tickets/e57f59e0-ffb7-11ed-9e92-33037837589c/none?models=0aed56a0-2d3f-11ee-8c0b-17a4bf167a31%2C825632b0-054a-11ee-ac5b-59e952832776)
image

If I group them by due date, I'm only seeing 2. The one with a year 2024 due date should be in 6+weeks category but it's not showing right now 
image

#### Grouping is not being remembered on refresh
If i select a grouping, then hit refresh. I'm back to `Group by: None`. Also I personally don't think template type and the group by should be part of the route, they should just be query strings like models - To me, the state of the route should present the page you're showing (i.e. tabular view), and the data within the page itself should be query strings (i.e. template/models selected, what am I grouping by).. but I'll leave this up to @sanmont3drepo 
img-19-52-23-02-52

#### Auto focus on title for new tickets doesn't happen if you've interacted with the side panel before

**Review by The-Daniel — CHANGES_REQUESTED** [review]

**Review by carmenfan — CHANGES_REQUESTED** [review]
### Need to disable the `Take screenshot` (or remove it, whatever is easiest) option when we're not in the viewer
image

**Review by carmenfan — DISMISSED** [review]

**Review by The-Daniel — CHANGES_REQUESTED** [review]

File: frontend/src/v5/ui/routes/dashboard/projects/tickets/ticketsTable/ticketsTableContent/ticketsTableGroup/ticketsTableGroup.component.tsx
Comment (carmenfan) [reply frontend/src/v5/ui/routes/dashboard/projects/tickets/ticketsTable/ticketsTableContent/ticketsTableGroup/ticketsTableGroup.component.tsx:1]: Yes in theory it should never be unset. But we shouldn't be hard coding for it (i.e. don't go and do if field == `priority` then get rid of the unset option). Observe the template, anything in the template that has a default value you'd never expect it to be unset.

Least of our worries though, if this is a lot of effort skip it and create a new issue.

File: frontend/src/v5/ui/routes/dashboard/projects/tickets/ticketsTable/ticketsTable.component.tsx
Comment (sanmont3drepo) [inline frontend/src/v5/ui/routes/dashboard/projects/tickets/ticketsTable/ticketsTable.component.tsx:81]: I think completed should be part of the url. To change that is as easy as changing this to use  useSearchParam instead of state, right? (in this case you could modify usesearhparam to receive a conversion func just to be suere is a boolean)

**Review by sanmont3drepo — CHANGES_REQUESTED** [review]

File: frontend/src/v5/ui/routes/routes.constants.ts
Comment (Amantini1997) [reply frontend/src/v5/ui/routes/routes.constants.ts:38]: We are in the side panel, some components such as GroupSettingsForm and CreateCommentBox, rely on `containerOrFederation` which is retrieved using `useParams`

Comment (Amantini1997) [reply frontend/src/v5/ui/routes/routes.constants.ts:38]: It is unfortunately not that simple, so, as discussed, I will keep it as is

File: frontend/src/v5/ui/routes/dashboard/projects/tickets/ticketsTable/ticketsTableContent/ticketsTableGroup/ticketsTableGroup.component.tsx
Comment (Amantini1997) [reply frontend/src/v5/ui/routes/dashboard/projects/tickets/ticketsTable/ticketsTableContent/ticketsTableGroup/ticketsTableGroup.component.tsx:1]: Will do in another issue, to not slow down the release

File: frontend/src/v5/ui/controls/sortedTableContext/sortedTableContext.tsx
Comment (The-Daniel) [inline frontend/src/v5/ui/controls/sortedTableContext/sortedTableContext.tsx:1]: This is minor, so can probably be made into a separate PR. Sorting assignees seems pretty arbitrary. I'm guessing we'd want to sort by the number of assignees?
image

Also Priority and Status should probably not sort alphabetically. 
image

**Review by The-Daniel — COMMENTED** [review]
I'm happy to approve this if we decide to turn this one thing into a new issue

File: frontend/src/v5/ui/controls/sortedTableContext/sortedTableContext.tsx
Comment (carmenfan) [reply frontend/src/v5/ui/controls/sortedTableContext/sortedTableContext.tsx:1]: yes create a new issue

**Review by The-Daniel — APPROVED** [review]

**Review by carmenfan — APPROVED** [review]

**Pull request merged**
